src/main/groovy/lang/NumberRange.java

```diff
@@ -0,0 +1,603 @@
+        }
+
+        // Performance detail:
+        // not using get(fromIndex), get(toIndex) in the following to avoid stepping over elements twice
+        final Iterator<Comparable> iter = new StepIterator(this, stepSize);
+
+        Comparable value = iter.next();
+        int i = 0;
+        for (; i < fromIndex; i++) {
+            if (!iter.hasNext()) {
+                throw new IndexOutOfBoundsException("Index: " + i + " is too big for range: " + this);
+            }
+            value = iter.next();
+        }
+        final Comparable fromValue = value;
+        for (; i < toIndex - 1; i++) {
+            if (!iter.hasNext()) {
+                throw new IndexOutOfBoundsException("Index: " + i + " is too big for range: " + this);
+            }
+            value = iter.next();
+        }
+        final Comparable toValue = value;
+
+        return new NumberRange(comparableNumber(fromValue), comparableNumber(toValue), comparableNumber(stepSize), true);
+    }
+
+    @Override
+    public String toString() {
+        return getToString(to.toString(), from.toString());
+    }
+
+    @Override
+    public String inspect() {
+        return getToString(InvokerHelper.inspect(to), InvokerHelper.inspect(from));
+    }
+
+    private String getToString(String toText, String fromText) {
+        String sep = inclusive ? ".." : "..<";
+        String base = reverse ? "" + toText + sep + fromText : "" + fromText + sep + toText;
+        return Integer.valueOf(1).equals(stepSize) ? base : base + ".by(" + stepSize + ")";
+    }
+
+    /**
+     * iterates over all values and returns true if one value matches.
+     * Also see containsWithinBounds.
+     */
+    @Override
+    public boolean contains(Object value) {
+        final Iterator it = new StepIterator(this, stepSize);
+        if (value == null) {
+            return false;
+        }
+        while (it.hasNext()) {
+            try {
+                if (compareEqual(value, it.next())) {
+                    return true;
+                }
+            } catch (ClassCastException e) {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void step(int numSteps, Closure closure) {
+        if (numSteps == 0 && compareTo(from, to) == 0) {
+            return; // from == to and step == 0, nothing to do, so return
+        }
+        final StepIterator iter = new StepIterator(this, multiply(numSteps, stepSize));
+        while (iter.hasNext()) {
+            closure.call(iter.next());
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Iterator<Comparable> iterator() {
+        // non thread-safe iterator
+        final Iterator<Comparable> innerIterator = new StepIterator(this, stepSize);
+        return new Iterator<Comparable>() {
+            @Override
+            public synchronized boolean hasNext() {
```

Comment:
I figured it was carried over from ObjectRange.  I just couldn't figure out why you'd want 2 or more threads to share an iterator.  Figured the typical use would be in a foreach loop and each thread would get their own instance of an iterator.   Otherwise, if sharing an instance a call to `hasNext()` followed by `next()` seems like it could return `null` for one of the threads after another thread got the last Number.